Java 第4回課題の提出

SuzukiNaoto0422 opened the pull request:
①CafeMenuでは文字列、値の要素もを持つリストを作成し書くリストに対しStremAPIを使用し要素の取得
②Studentクラスの作成
③Studentクラスを使用したStudentListクラスの中でリストを作成しStremAPIを使用し要素の取得

Commit:
削除

Commit:
リストを作成、StremAPIを使用し要素の取得

Commit:
名前、身長、血液型のフィールドを持つStudentクラスの作成

Commit:
Studentクラスを利用しリストの作成、StremAPIを使用し要素の抽出

yoshi-koyama commented on src/CafeMenu.java:
```diff
@@ -0,0 +1,48 @@
+        System.out.println(highPrice);
+
+        option.stream()
+                .map(String::toUpperCase) //toUpperCase()で大文字に
```
メソッド参照もつかえていてgood 👍

yoshi-koyama commented on src/CafeMenu.java:
```diff
@@ -0,0 +1,48 @@
+
+        //minとmaxの引数は、比較に用いられるインターフェースのComparator、戻り値は、Optionalを用いる。
+        Optional<Integer> minPrice  = priceLists.stream().min(Comparator.naturalOrder());//最小の値の取得
+        System.out.println(minPrice.get());
```
`.get`を使うと万が一値が存在しない場合にNoSuchElementExceptionが発生します。
Optionalは値がない時の動作を指定するためのものなので`get`は使わないようにしましょう。
（getなどという明らかに使っても大丈夫そうな名前のメソッドを作ってしまったことはOracleのJava Language ArchitectのBrianさん※もおっしゃっていました。
※Java言語界隈の偉い人

https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type
> In retrospect, we should have called get something like getOrElseThrowNoSuchElementException or something that made it far clearer that this was a highly dangerous method that undermined the whole purpose of Optional in the first place. Lesson learned.

yoshi-koyama commented on src/Student.java:
```diff
@@ -0,0 +1,26 @@
+        this.height = height;
+        this.bloodType = bloodType;
+    }
+
+
+    public String getName() {return name;}
```
空白行1つで良いと思います。

yoshi-koyama commented on src/Student.java:
```diff
@@ -0,0 +1,26 @@
+
+    //「Student」クラスを表示するときに、「name」フィールドや「height」フィールドの値を知りたい場合は、「toString」メソッドをオーバーライドして、それらの値を含む文字列を返すようにする。
+    //「toString」メソッドは、オブジェクトを文字列に変換するメソッド。Javaでは、オブジェクトを直接出力する場合、「toString」メソッドが呼び出され、その戻り値が出力。
+        @Override
+        public String toString() {
```
toStringメソッドのインデントが気になりました。
https://github.com/SuzukiNaoto0422/Java_4/blob/456f6b2a8c738613218c7b7c80596ab1c6d1be5e/src/Student.java#L17-L23

↑を見てもらうとわかりますが、getterやコンストラクタとインデントがずれていますね。

yoshi-koyama commented on src/StudentList.java:
```diff
@@ -0,0 +1,37 @@
+        System.out.println(heightList);
+
+        //血液型でグルーピング
+        Map<String,List<Student>> grp = list.stream()
```
Mapの変数名はわかりにくくなりがちなのですが、個人的には
keyToValue
か
valueByKey
のような形式が良いと思っています。
```suggestion
        Map<String, List<Student>> bloodTypeToStudents = list.stream()
```

Commit:
第4回課題訂正
・CafeMenuクラスのminPrice、maxPriceの訂正
・Studentクラスのインデント修正
・StudentListクラスのMapの変数名の訂正

yoshi-koyama commented on src/CafeMenu.java:
```diff
@@ -0,0 +1,48 @@
+                .forEach(System.out::println); //forEachでストリーム内の要素をそれぞれ取り出し、System.out.printlnを使用して標準出力に出力。
+
+        //Optional<T>.orElse(T other)メソッドを使用することで、値が存在しない場合にデフォルト値を返すことができる。
+        Integer minPrice  = priceLists.stream().min(Comparator.naturalOrder()).orElse(-1);//最小の値の取得
```
最小値のデフォルトなら0が適当だと思います。

yoshi-koyama commented on src/StudentList.java:
```diff
@@ -0,0 +1,37 @@
+        Map<String, List<Student>> bloodTypeToStudents = list.stream() //グループの変数名はkeyToValue かvalueByKeyのような形式が良い
+                .collect(Collectors.groupingBy(Student::getBloodType));
+
+        //grp.entrySet()を使用して、Mapオブジェクトのエントリを取得
```
ここのコメントがgrpのままですね。

yoshi-koyama commented on src/Student.java:
```diff
@@ -0,0 +1,25 @@
+
+    //「Student」クラスを表示するときに、「name」フィールドや「height」フィールドの値を知りたい場合は、「toString」メソッドをオーバーライドして、それらの値を含む文字列を返すようにする。
+    //「toString」メソッドは、オブジェクトを文字列に変換するメソッド。Javaでは、オブジェクトを直接出力する場合、「toString」メソッドが呼び出され、その戻り値が出力。
+        @Override
```
インデント直ってない？

Commit:
第4回課題訂正(2回目）
・CafeMenuクラスのminPrice、maxPriceの訂正
・Studentクラスのインデント修正
・StudentListクラスのコメントの訂正
